input/sdl_gamepad.c

```diff
@@ -0,0 +1,244 @@
+        MP_ERR(src, "SDL_Init failed\n");
+        mp_input_src_init_done(src);
+        return;
+    };
+
+    SDL_GameController *controller;
+    struct priv *p = talloc_zero(src, struct priv);
+    char guid[GUID_LEN];
+
+    if (SDL_NumJoysticks() <= 0) {
+        MP_VERBOSE(src, "no joysticks found");
+        mp_input_src_init_done(src);
+        return;
+    }
+
+    MP_VERBOSE(src, "connected controllers: %i\n", SDL_NumJoysticks());
+
+    for (int i = 0; i < SDL_NumJoysticks(); ++i) {
+        if (SDL_IsGameController(i)) {
+            controller = SDL_GameControllerOpen(i);
+            SDL_JoystickID id = SDL_JoystickInstanceID(
+                SDL_GameControllerGetJoystick(controller));
+
+            SDL_JoystickGetGUIDString(
+                SDL_JoystickGetGUID(SDL_GameControllerGetJoystick(controller)),
+                guid, GUID_LEN);
+
+            if (controller) {
+                MP_VERBOSE(
+                    src, "detected controller #%i: %s, guid: %s\n",
+                    id, SDL_GameControllerName(controller), guid);
+
+                // stop at first controller apparently SDL can't open more
+                // than one controller anyway?
+                break;
+            }
+        }
+    }
+
+    atomic_store(&p->cancel_requested, false);
+    src->priv = p;
+    src->cancel = request_cancel;
+    src->uninit = uninit;
+    mp_input_src_init_done(src);
+
+    SDL_Event ev;
+
+    while (!atomic_load(&p->cancel_requested)) {
+        while (SDL_PollEvent(&ev) != 0) {
+            switch (ev.type) {
+                case SDL_CONTROLLERBUTTONDOWN: {
+                    const int key = lookup_button_mp_key(ev.cbutton.button);
+                    if (key != INVALID_KEY) {
+                        mp_input_put_key(src->input_ctx, key | MP_KEY_STATE_DOWN);
+                    }
+                    continue;
+                }
+                case SDL_CONTROLLERBUTTONUP: {
+                    const int key = lookup_button_mp_key(ev.cbutton.button);
+                    if (key != INVALID_KEY) {
+                        mp_input_put_key(src->input_ctx, key | MP_KEY_STATE_UP);
+                    }
+                    continue;
+                }
+                case SDL_CONTROLLERAXISMOTION: {
+                    const int key =
+                        lookup_analog_mp_key(ev.caxis.axis, ev.caxis.value);
+                    if (key != INVALID_KEY) {
+                        mp_input_put_key(src->input_ctx, key);
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+
+    if (controller != NULL) {
+        SDL_GameControllerClose(controller);
+    }
+
+    // must be called on the same thread of SDL_InitSubSystem, so uninit
+    // callback can't be used for this
+    SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+}
+
+void mp_input_sdl_gamepad_add(struct input_ctx *ictx)
+{
+    mp_input_add_thread_src(ictx, NULL, read_gamepad_thread);
```

Comment:
I planned to kill this API, only the win32 pipe stuff still used it. I may request that you adjust this code some time in the future when it comes to it.

Reply:
Ok, I actually use this feature.